### ISSUE TITLE
Umineko patch "options"

### ISSUE THREAD

**Issue by drojf** [07th-mod/umineko-question#11]
# Umineko Options (for V1.0)

I saw some questions on having various combinations of sprites/backgrounds/voices, this should cover all of them. Let me know if the following work for you (in other words, **these may not work and have not been tested yet!**).

### Important

Please read the information in the latest release before applying any of these options (specifically, save game compatibility). This post is over 6 months old, and may contain outdated information:

https://github.com/07th-mod/umineko-question/releases/latest

### Background Information

The game will prefer files on disk to files in .nsa files, so you don't need to re-archive the .nsa files after you've extracted them. Just make sure they match the file structure shown below.

Script file: in the root directory, named `0.utf`
Voices: located in `/voice` (not part of the original steam release)
'Big' Character Sprites: located in `/big/bmp/tati` folder
Backgrounds/effects: all over the place. best to just define backgrounds/effect files as "all images, minus the sprites"
The actual backgrounds are in `/bmp/background`

'Small' Character Sprites: located in `/bmp/tati`. Only shown if enabled, otherwise never used!

NSA Extraction tool: download from http://unclemion.com/onscripter/releases/ under the "ONScripter Tools" section

What happens in 16:9 mode and a 4:3 image is shown? The image will be drawn centered, in the middle of the screen. 

---
### Disable voices (can be combined with any of the below)

If you don't install voice files, the game will just not play any voices

### PS3 Voices only, no graphical upgrades (_confirmed_ in https://github.com/07th-mod/umineko-question/issues/1#issuecomment-248576028)

Summary: use old version of script which has voices but not the graphical patches
- download the voice files, make sure they are copied to `umineko/voice`
- **EDIT:** There is a new version of the "voice only" patch, see https://github.com/07th-mod/umineko-question/blob/master/InDevelopment/ManualUpdatesVoiceOnly/0.utf
- ~~download the old script https://github.com/07th-mod/umineko-question/commit/ff5cd239c514fb33ca610c79ef76ab3fcf1d6beb (last version of 0.utf before we started messing with the backgrounds)~~
~~- It's advisable to change the following lines in the above script:~~
~~`umalias default_vol, 70`~~

~~change the above line to:~~

~~`umalias default_vol, 15`~~
~~`voicevol 20`~~

~~as shown being changed here: https://github.com/07th-mod/umineko-question/commit/dcfa55ffbc9831dc07d00a8dd3098d74e27dd1a9 (version with music/voice volume adjusted)~~

~~This sets the music volume, and the voice volume (set it how you wish).~~
~~- do NOT patch anything else. The game will happily load voice files by script modification only.~~

~~Note: because you are using the old version of the script, it will have the 'achievement unlocks early' bug, and some small typos which MG have now patched, see https://github.com/07th-mod/umineko-question/commit/1788fcb67cdf6df981eeb6e18a13de185c80da12 (spoilers I guess, since this shows which lines were changed).~~
### PS3 Sprites only

Summary: put new sprites in Umineko folder so the game prefers them to the original sprites
- extract the graphics.nsa
- copy out ONLY the sprites folder, and put it in the umineko folder (eg structure as `umineko/big/bmp/tati`)
- place in the umineko folder
- game will prefer the sprites in the filesystem to the ones in the steam arc.nsa (hopefully)
- do not patch anything else (do not patch exe, do not patch 0.utf). This ensures the game renders at 16:9 correctly
### PS3 Backgrounds only

Summary: delete sprite folder from graphics.nsa patch, then proceed as normal
- extract the graphics.nsa
- delete the sprites folder
- put the remaining folders the umineko root directory (either copy directly in or re-archive)
- carry out all the remaining installation steps as normal
### Text shown at bottom of the screen instead of covering whole screen
- currently not implemented, but if anyone wants to give it a go, it's pretty easy to do (I think).
- read https://github.com/07th-mod/umineko-question/issues/1#issuecomment-235526697 and http://unclemion.com/onscripter/api/NScrAPI.html#setwindow (it's the first function, not the second function)
- you will need to modify all the 'setwindow' functions in the script, EXCEPT for the ones which occur at 'special events' (preferably by some sort of patching script, [similar to this](https://github.com/07th-mod/umineko-question/blob/master/widescreen/patchUminekoToWidescreen.py#L136), instead of doing it manually)  
- I believe the original 'Umitweak' people did the same thing - may be worth contacting them or downloading their script to see how they did it. 
- Let us know so we can add it to the next release :) .
### drojf's special voice player
- if you have the original voice files from the ps3 game (.at3 format), you can use my special loader to play them. 
- voice quality improvement is marginal/small (less compression artifacts due to double encoding)
- if at least 2 people ask me for the original voice files I'll upload the raw voice files so the loader can be used (I only found one place hosting the voice files, and some voices were missing!)
- see https://github.com/07th-mod/umineko-question/tree/master/umihook for more details

**Comment by ReitoKanzaki** [comment]
Would definitely like to have the original voices as at least an option.

**Comment by drojf** [comment]
You mean PS3 voices only? or something else?

**Comment by ReitoKanzaki** [comment]
Sorry, meant "original Atrac3+ PS3 voices".

**Comment by marakeusz** [comment]
> umalias default_vol, 15
voicevol 20

I wonder why **default_vol** is lowered. Setting **voicevol** to etc. 75 would be a bad thing? I am asking because in the script volumes are set in various ways and **default_vol** is not used in many cases.

Modifying just **default_vol** seems a little messy. There are lot of situations in which bgm vol is set to default but environment volumes (etc. rain) are set to particular values - often much higher (closer to old default_vol). Results: we cannot hear voices, we cannot hear music, we can hear loud rain - very bad experience. At first I just removed the mod but after while I edited the default_vol to original value and I set the voicevol to 75. It is pretty good now.

Is that make a sense to you? I am new to nscripter/ponscripter.

**Comment by drojf** [comment]
Yes that makes sense.

Initially, we didn't change the default_vol variable. We had some issues when if the music/voice(/sfx) played at the same time with default volume (70) the audio would clip. To be honest I haven't played through the game enough to notice if there were any issues due to the game itself changing the volume. 

If you can play through the game with your settings and don't notice any audio clipping issues, please report back - we'll make that the default values.

The final option is my voice player hook, which plays the audio files outside the game (it detects when voice files are attempted to be loaded, then plays the corresponding .at3 file). I made this because I was annoyed that the .at3 files (which are already low bitrate) had to be encoded again into .ogg format. You can just hear the artifacting on certain voice files, but it's minor, so I decided not to try and put it in the modded game (we had enough trouble just putting out the patch in the first place lol). 

The up-side of my audio player is that you can set the volume for the voices and the game won't interfere. It also appears to prevent clipping somehow (probably the windows mixer prevents it). It would also fix the layered voice volume bug too (once I add the feature so it can play multiple files at once).

**Comment by ItaloKnox** [comment]
The low default volumes are mostly my fault here. I'm running the game in a very loud setup (I work with music), so most likely I tuned it down a little too much here. Let me know how your game sounds like after changing these values, I will fix it asap.

**Comment by marakeusz** [comment]
I just finished the 1st episode so a lot of content ahead of me but I cannot say I remember any clipping.

**Comment by marakeusz** [comment]
**Text shown at bottom of the screen instead of covering whole screen**
> you will need to modify all the 'setwindow' functions in the script, EXCEPT for the ones which occur at 'special events'

That option sounds interested. I am curious what do you mean by **special events** Can you provide example?

**Comment by drojf** [comment]
the setwindow is called 58 times in the script, most of the time with the same arguments:

`setwindow 31,16,25,20,23,23,0,2,20,1,1,#999999,0,0,639,479  //original`

I modified all setwindow functions by scaling the values so it would match widescreen

`setwindow 137,16,25,20,23,23,0,2,20,1,1,#999999,0,0,853,479 //widescreen version` 

So to make it a smaller text box, you simply change the bounds of the box, and reduce the font size [(and maybe some other things are required)](http://unclemion.com/onscripter/api/NScrAPI.html#setwindow).

However sometimes `setwindow` is called with different arguments to normal:

```
setwindow 137,16,22,16,50,50,0,2,20,1,1,#999999,0,0,853,479
lsph %tmp,":s#00ffff　　　　　ぺったん　　　ぺったん　　　胸ぺったん☆",0,0 : inc %tmp
setwindow 137,16,22,16,25,25,0,2,20,1,1,#999999,0,0,853,479
;lsph %tmp,":s#00ffff☆　　（　＾∇＾）σ）≧∇≦）　　☆",0,0 : inc %tmp
````

This is the "nicovideo comments on screen" and lyrics, and therefore should be left as is. This happens when jessica/kannon is at the cultural festival. I'd expect there to be other special events in the game (eg when text is shown in the center of the screen for emphasis). Those cases should also be left alone.

Now that I look at it, there are probably only a few places in the script where you'd have to actually change the `setwindow` function, that is at the start of each chapter. But it'd probably be best to go through each `setwindow` and verify what it is for, in case it's used inside a function or something like that.

I just used notepad++'s "find all in current document" to find all instances of it, [which looks like this](https://gist.github.com/drojf/efca65b0969cbf79250a80196e123ac3). From there, you can click on each result to jump to it and inspect what the 'setwindow' is used for.

Testing is pretty easy - change the setwindow you think applies for each chapter, then skip through the chapter. The window should never reset itself to fill the whole screen (except for special events).

Personally I would go through the script, and add a comment at the end of each setwindow function if I know what it is for (eg 'start of ep1', 'cultural festival music', 'end of cultural festival'). Then I could identify how many setwindow functions are 'unknown', and would help catch cases where the window is reset unexpectedly in the middle of an episode.

**Comment by marakeusz** [comment]
I thought this through and I think it would more annoying than useful. I would not enjoy reading whole novel at the bottom.

Maybe if only the dialogue lines would be on the bottom and rest of the text would be displayed in the "normal way". It is not impossible... Still not perfect.

Anyway - thanks for your help - good to know.

**Comment by yagamisempai** [comment]
hello, i want to have voices only so i downloaded the voice archives and put them on the umineko folder, i go to https://github.com/07th-mod/umineko-question/blob/master/InDevelopment/ManualUpdatesVoiceOnly/0.utf but it seems i cant download the archive so what i did was to create a 0.utf file copying the text but my game crash, what can i do? i only want voices thank you.

**Comment by ItaloKnox** [comment]
@yagamisempai Right click on [this link](https://github.com/07th-mod/umineko-question/raw/master/InDevelopment/ManualUpdatesVoiceOnly/0.utf) and use the option "save as...", then put the name as 0.utf (if it's not there already)

**Comment by drojf** [comment]
All of the options described here have been implemented already except for:

- "original Atrac3+ PS3 voices" - It is unlikely we will support the original ps3 voices (since the voice quality improvement is very little), and there hasn't been much work put into modifying the engine.

- PS3 sprites only - This option isn't requested very often. For question arcs, you would just need to resize the background images. For the answer arcs, you would need to re-map the background images to suit PS3 filenames, then resize them too. But I don't think we will implement this since not many people ask for it.

The "PS3 Backgrounds only" is an odd one - technically we kind of support it by doing this:

- install ps3 patch with "ryukishi sprites" option
- delete the voices folder to remove voices

I don't think many people would want this option though. I'm not sure why I listed it here.

As we are unlikely to support the two remaining options, I'm closing this issue - any other requests should be opened as a new issue.
